### PR TITLE
Liberty view for IO cells update. 

### PR DESCRIPTION
**Pull request by KrzysztofHerman**
The updated view contains definition of all IO cells provided in the library hence it is feasible to use it in ORFS to design a pad ring.

### CODE REVIEW

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 1]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
+                    );
+                }
+                fall_transition (delay_template_2x2) {
+                    values ( \
+                        "200, 200", \
+                        "200, 200" \
+                    );
+                }
+            }
+                }pin (pad) {
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]:
```suggestion
            }
            pin (pad) {
```

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 2]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
+                    );
+                }
+                fall_transition (delay_template_2x2) {
+                    values ( \
+                        "200, 200", \
+                        "200, 200" \
+                    );
+                }
+            }
+                }    }
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]:
```suggestion
         }
 }
```

Comment (dnltz) [reply ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: See all following cells have the same issue.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 3]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
+
+    cell (sg13g2_IOPadIOVss) {
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib:1440]: Can you move both IO cells at the beginning or move Vss/Vdd cells from the beginning here?

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 4]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
+                    );
+                }
+                fall_transition (delay_template_2x2) {
+                    values ( \
+                        "200, 200", \
+                        "200, 200" \
+                    );
+                }
+            }
+                }pin (c2p) {
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: Should be next line.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 5]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: Should be next line.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 6]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: Should be next line.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 7]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: Should be next line.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 8]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: Should be next line.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 9]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: Should be next line.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 10]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: Should be next line.

File: ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib [hunk 11]
```diff
@@ -86,4 +86,2956 @@ library (sg13g2_io_dummy) {
...
+                    );
+                }
+                fall_transition (delay_template_2x2) {
+                    values ( \
+                        "200, 200", \
+                        "200, 200" \
+                    );
+                }
+            }
+            timing () {
```
Comment (dnltz) [inline ihp-sg13g2/libs.ref/sg13g2_io/liberty/sg13g2_io_dummy.lib]: I don't see any obvious issue here with these two timings. I wonder if timings are missing in the pin c2p and pin c2p_en definitions...

**Review by dnltz — COMMENTED** [review]
This file has some indentation issues.

**Comment by KrzysztofHerman** [comment]
There were some updates, which limit the cells to the list defined in `sg13g2_io.lef`:

```
MACRO sg13g2_Corner
MACRO sg13g2_Filler200
MACRO sg13g2_Filler400
MACRO sg13g2_Filler1000
MACRO sg13g2_Filler2000
MACRO sg13g2_Filler4000
MACRO sg13g2_Filler10000
MACRO sg13g2_IOPadIn
MACRO sg13g2_IOPadOut4mA
MACRO sg13g2_IOPadOut16mA
MACRO sg13g2_IOPadOut30mA
MACRO sg13g2_IOPadTriOut4mA
MACRO sg13g2_IOPadTriOut16mA
MACRO sg13g2_IOPadTriOut30mA
MACRO sg13g2_IOPadInOut4mA
MACRO sg13g2_IOPadInOut16mA
MACRO sg13g2_IOPadInOut30mA
MACRO sg13g2_IOPadAnalog
MACRO sg13g2_IOPadIOVss
MACRO sg13g2_IOPadIOVdd
MACRO sg13g2_IOPadVss
MACRO sg13g2_IOPadVdd

``` 
Also some properties of the filler cell were added together with a fix on positive/negative unate. 

I ran ORFS using these updated views and still got some issues that are listed below: 

```

logs/ihp-sg13g2/mini/base/6_1_fill.log:5:[WARNING FIN-0010] Skipping layer OVERLAP.
logs/ihp-sg13g2/mini/base/6_1_fill.log:6:[WARNING FIN-0010] Skipping layer LOCKED.
logs/ihp-sg13g2/mini/base/6_1_fill.log:7:[WARNING FIN-0010] Skipping layer LOCKED1.
logs/ihp-sg13g2/mini/base/6_1_fill.log:8:[WARNING FIN-0010] Skipping layer LOCKED2.
logs/ihp-sg13g2/mini/base/6_1_fill.log:9:[WARNING FIN-0010] Skipping layer GatPoly.
logs/ihp-sg13g2/mini/base/6_1_fill.log:10:[WARNING FIN-0010] Skipping layer Cont.
logs/ihp-sg13g2/mini/base/6_1_fill.log:14:[WARNING FIN-0010] Skipping layer Via1.
logs/ihp-sg13g2/mini/base/6_1_fill.log:18:[WARNING FIN-0010] Skipping layer Via2.
logs/ihp-sg13g2/mini/base/6_1_fill.log:22:[WARNING FIN-0010] Skipping layer Via3.
logs/ihp-sg13g2/mini/base/6_1_fill.log:26:[WARNING FIN-0010] Skipping layer Via4.
logs/ihp-sg13g2/mini/base/6_1_fill.log:30:[WARNING FIN-0010] Skipping layer TopVia1.
logs/ihp-sg13g2/mini/base/6_1_fill.log:34:[WARNING FIN-0010] Skipping layer TopVia2.
logs/ihp-sg13g2/mini/base/5_3_route.log:8:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer Cont
logs/ihp-sg13g2/mini/base/5_3_route.log:9:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer Via1
logs/ihp-sg13g2/mini/base/5_3_route.log:10:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer Via1
logs/ihp-sg13g2/mini/base/5_3_route.log:11:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer Via2
logs/ihp-sg13g2/mini/base/5_3_route.log:12:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer Via3
logs/ihp-sg13g2/mini/base/5_3_route.log:13:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer Via4
logs/ihp-sg13g2/mini/base/5_3_route.log:14:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer TopVia1
logs/ihp-sg13g2/mini/base/5_3_route.log:15:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer TopVia1
logs/ihp-sg13g2/mini/base/5_3_route.log:16:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer TopVia2
logs/ihp-sg13g2/mini/base/5_3_route.log:17:[WARNING DRT-0349] LEF58_ENCLOSURE with no CUTCLASS is not supported. Skipping for layer TopVia2
logs/ihp-sg13g2/mini/base/2_1_floorplan.log:27:[WARNING PAD-0033] Could not find a block terminal associated with net: "IO_CORNER_NORTH_WEST_INST.iovdd_RING", creating now.
logs/ihp-sg13g2/mini/base/2_1_floorplan.log:29:[WARNING PAD-0033] Could not find a block terminal associated with net: "IO_CORNER_NORTH_WEST_INST.iovss_RING", creating now.
logs/ihp-sg13g2/mini/base/2_1_floorplan.log:31:[WARNING PAD-0033] Could not find a block terminal associated with net: "IO_CORNER_NORTH_WEST_INST.vss_RING", creating now.
logs/ihp-sg13g2/mini/base/2_1_floorplan.log:33:[WARNING PAD-0033] Could not find a block terminal associated with net: "IO_CORNER_NORTH_WEST_INST.vdd_RING", creating now.

```
Maybe @stafverhaegen-chipflow could provide some solutions for `DRT-0394` and `PAD-0033` ? I guess the first can be solved downgrading the LEF view to 5.7.

**Comment by FatsieFS** [comment]
I can change version to 5.7 if that is confirmed to solve the problem.
For the `PAD-0033` it is not clear to me what is wrong with the LEF so will need some guidance. When I look to similarity between Filler cells and corner cells I can't see what is wrong.
See also snapshot of corner when imported in KLayout:
image

**Comment by dnltz** [comment]
Could it be missing pin definitions for power/ground? All IOPad cells have following pg_pins but not Corner/Filler cells.

```
        pg_pin (vss) {
            pg_type : primary_ground;
            voltage_name : "vss";
        }
        pg_pin (vdd) {
            pg_type : primary_power;
            voltage_name : "vdd";
        }
        pg_pin (iovss) {
            pg_type : primary_ground;
            voltage_name : "iovss";
        }
        pg_pin (iovdd) {
            pg_type : primary_power;
            voltage_name : "iovdd";
        }
```

**Comment by KrzysztofHerman** [comment]
@FatsieFS feel free to provide downgraded version of the LEF. As for `PAD-0033` it is strange that it complains only on North_West corner.

**Pull request merged**